src/operations/extension/reduction_partial_rows.hpp

```diff
@@ -0,0 +1,237 @@
+#include <CL/sycl.hpp>
+#include <string>
+
+namespace blas {
+
+template <typename operator_t, typename input_t, typename output_t, int ClSize,
+          int WgSize, int WorkPerItem, typename element_t>
+class ReductionPartialRows {
+ public:
+  using index_t = typename input_t::index_t;
+  using value_t = element_t;
+
+  /* Reading some compile-time parameters from a structure.
+   * See the header file for the definition of this structure */
+  using params_t =
+      ReductionRows_Params<index_t, element_t, ClSize, WgSize, WorkPerItem>;
+  static constexpr index_t cl_elems = params_t::cl_elems;
+  static constexpr index_t rows_per_item = params_t::rows_per_item;
+  static constexpr index_t work_group_size = params_t::work_group_size;
+  static constexpr index_t work_group_rows = params_t::work_group_rows;
+  static constexpr index_t work_group_cols = params_t::work_group_cols;
+  static constexpr index_t local_memory_rows = params_t::local_memory_rows;
+  static constexpr index_t local_memory_cols = params_t::local_memory_cols;
+  static constexpr index_t local_memory_size = params_t::local_memory_size;
+
+  /* Neutral value for this reduction operator */
+  static constexpr value_t init_val = operator_t::template init<output_t>();
+
+  /* Input and output buffers */
+  input_t in_;
+  output_t out_;
+
+  /* Matrix dimensions */
+  const index_t rows_;
+  const index_t cols_;
+  const index_t leading_dim_;
+
+  /* Work groups per dimension */
+  const index_t group_count_rows_;
+  const index_t group_count_cols_;
+
+  SYCL_BLAS_INLINE ReductionPartialRows(input_t in, output_t out,
+                                        index_t group_count_cols)
+      : in_(in),
+        out_(out),
+        rows_(in_.get_size_row()),
+        cols_(in_.get_size_col()),
+        leading_dim_(in_.getSizeL()),
+        group_count_rows_((rows_ - 1) / local_memory_rows + 1),
+        group_count_cols_(group_count_cols) {}
+
+  void bind(cl::sycl::handler &h) {
+    in_.bind(h);
+    out_.bind(h);
+  }
+  void adjust_access_displacement() {
+    in_.adjust_access_displacement();
+    out_.adjust_access_displacement();
+  }
+  SYCL_BLAS_INLINE bool valid_thread(cl::sycl::nd_item<1> ndItem) const {
+    return true;
+  }
+
+  /*!
+   * @brief get_workgroup_cluster. This function is used to find the optimum
+   * number of work groups required to execute each partial reduction step.
+   */
+  SYCL_BLAS_INLINE index_t get_workgroup_cluster() noexcept {
+    return ((rows_ - 1) / local_memory_rows + 1) * group_count_cols_;
+  }
+  /*!
+   * @brief get_num_workgroup_cluster. This function is used to extend the
+   * number of work group clusters, in order to make sure that at least 4
+   * operations are available per work group. The number 4 is used based on
+   * empirical research.
+   */
+  SYCL_BLAS_INLINE index_t
+  get_num_workgroup_cluster(index_t compute_units) noexcept {
+    return ((4 * compute_units - 1) / get_workgroup_cluster() + 1);
```

Comment:
Put 4 in a constant somewhere.

Reply:
This function is exactly copied from the legacy Gemm algorithm (and the constant verified empirically as the most performant).
I agree that the constant could be put elsewhere to be controlled (or even made a template parameter), but where would you put it? Unless there's a common location that makes sense, the best place for readability is right here

Reply:
At the moment at least, I'd just put it at the top of the file.

Reply:
In that case, I'd rather let it here, when you're reading the code you don't have to jump somewhere else to get the value. Moreover it's not meant to be changed often, or even to be changed at all.

Reply:
Why do you need to know the value?

Reply:
Because it gives you an idea of the number of work groups. And I think it's better to let this value here at the moment, unless we want to create a parameter to control this (e.g to use different values on different architectures).

Reply:
(this function can be seen as a heuristics to select the number of work groups, which is why I think it's fine to have the constant here; and the constant is not meant to be used anywhere else)

Reply:
Ah, so in a way the whole "function" could be seen as the constant then? That's fair then.

Reply:
I have addressed the issue with a constexpr variable at the moment. We can later think about making it a template parameter if we want per-platform control.